src/Microsoft.AspNet.Server.Kestrel/Filter/StreamSocketOutput.cs

```diff
@@ -62,18 +84,49 @@ public MemoryPoolIterator2 ProducingStart()
 
         public void ProducingComplete(MemoryPoolIterator2 end)
         {
-            var block = _producingBlock;
-            while (block != end.Block)
+            lock (_writeLock)
             {
-                _outputStream.Write(block.Data.Array, block.Data.Offset, block.Data.Count);
+                ProducingCompleteAsync(ref end)
+                    .ContinueWith(_producingCompleteError,
+                        _logger,
+                        TaskContinuationOptions.OnlyOnFaulted);
+            }
+        }
+
+        private Task ProducingCompleteAsync(ref MemoryPoolIterator2 end)
+        {
+            var task = TaskUtilities.CompletedTask;
+            MemoryPoolBlock2 block;
+            try
+            {
+                // When the Async calls return the tasks they have copied the data to the backing buffer
+                // also in writing in a lock so is safe to send one after another return the final task
+
+                block = _producingBlock;
+                while (block != end.Block)
+                {
+                    _outputStream.WriteAsync(block.Data.Array, block.Data.Offset, block.Data.Count, CancellationToken.None);
```

Comment:
...unless it was only `immediate:true` when `Flush` is called - but that's probably just asking for trouble...